### PR TITLE
feat: templates for network and storage configurations

### LINKED ISSUE #262: Add support for builds using static IP address assignment
### Code of Conduct

- [X] I have read and agree to the project's [Code of Conduct](https://github.com/vmware-samples/packer-examples-for-vsphere/blob/main/CODE_OF_CONDUCT.md).

### Community Note
- Vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue initial description to help the maintainers prioritize.
- Do not leave "+1" or other comments that do not add relevant information or questions.
- If you are interested in working on this issue or have submitted a pull request, please leave a comment.

### Description

Add support for builds using static IP address assignment.

### Use Case(s)

Provides examples for environments that are unable to use DHCP (default).

### Potential Configuration

To be determined.

There's not a native method in the [Packer Builder for VMware vSphere](https://www.packer.io/docs/builders/vsphere/vsphere-iso) to use static IP address assignment. 

Options to enable this may include:

#### Examples

**Ubuntu 18.04**

Update the ks.cfg preseed file.

```vim
# Networking
d-i netcfg/disable_autoconfig boolean true

# IPv4 Setup
d-i netcfg/get_ipaddress string <static_ipv4_address>
d-i netcfg/get_netmask string <static_ipv4_netmask>
d-i netcfg/get_gateway string <static_ipv4_gateways>
d-i netcfg/get_nameservers string <static_ipv4_dnsservers>
d-i netcfg/confirm_static boolean true
```

**Ubuntu 20.04 and later**

Update the `user-data` used by cloud-init.

```yaml
  network:
    network:
      version: 2
      ethernets:
        ens33:
         addresses: <static_ipv4_address_cidr>
         gateway4: <static_ipv4_gateway>
         nameservers:
           addresses: [<static_ipv4_dnsservers>]

```

Other Linux distributions will vary.

**Windows Server / Desktop**

Update the `autounattend.pkrtpl.hcl` with to set the IP address after the installation of VMware Tools.

```xml
         <FirstLogonCommands>
            <SynchronousCommand wcm:action="add">
               <CommandLine>%SystemRoot%\system32\WindowsPowerShell\v1.0\powershell.exe -Command "Set-ExecutionPolicy -ExecutionPolicy RemoteSigned -Force"</CommandLine>
               <Description>Set Execution Policy 64-Bit</Description>
               <Order>1</Order>
               <RequiresUserInput>true</RequiresUserInput>
            </SynchronousCommand>
            <SynchronousCommand wcm:action="add">
               <CommandLine>%SystemRoot%\system32\WindowsPowerShell\v1.0\powershell.exe -Command "Set-ExecutionPolicy -ExecutionPolicy RemoteSigned -Force"</CommandLine>
               <Description>Set Execution Policy 32-Bit</Description>
               <Order>2</Order>
               <RequiresUserInput>true</RequiresUserInput>
            </SynchronousCommand>
            <SynchronousCommand wcm:action="add">
               <CommandLine>%SystemRoot%\system32\WindowsPowerShell\v1.0\powershell.exe -File F:\windows-vmtools.ps1</CommandLine>
               <Order>3</Order>
               <Description>Install VMware Tools</Description>
            </SynchronousCommand>
            <SynchronousCommand wcm:action="add">
               <CommandLine>%SystemRoot%\system32\WindowsPowerShell\v1.0\powershell.exe -File F:\windows-init.ps1</CommandLine>
               <Order>4</Order>
               <Description>Initial Configuration</Description>
            </SynchronousCommand>
             <SynchronousCommand wcm:action="add">
                 <CommandLine>cmd.exe /c netsh interface ip set address "Ethernet0" static <static_ipv4_address> <static_ipv4_netmask> <static_ipv4_gateway></CommandLine>
                 <Description>Assign IP</Description>
                 <Order>5</Order>
                 <RequiresUserInput>true</RequiresUserInput>
             </SynchronousCommand>
             <SynchronousCommand wcm:action="add">
                 <CommandLine>cmd.exe /c netsh interface ip add dns "Ethernet0" <static_ipv4_dnsservers></CommandLine>
                 <Description>Assign DNS</Description>
                 <Order>6</Order>
                 <RequiresUserInput>true</RequiresUserInput>
         </FirstLogonCommands>

```

### References

None.

### PR DESCRIPTION
**Pull request by sestegra**
# 

## Summary of Pull Request

This PR implements some templates with unit tests for managing custom network and storage configurations.
The integration with the current `builds` tree should be discussed.

I suggest creating an separate files for network and storage configuration to ease the customization.

For instance, for Ubuntu 22.04 build, the content of the folder will be the following:
```
builds/linux/ubuntu/22-04-lts
├── data
│   ├── meta-data
│   ├── network.pkrtpl.hcl
│   ├── storage.pkrtpl.hcl
│   └── user-data.pkrtpl.hcl
├── linux-ubuntu.auto.pkrvars.hcl
├── linux-ubuntu.pkr.hcl
├── variables-network.pkr.hcl
├── variables-storage.pkr.hcl
└── variables.pkr.hcl
```

The Packer configuration will contain something like this:
```hcl
data_source_content = {
  "/meta-data" = file("${abspath(path.root)}/data/meta-data")
  "/user-data" = templatefile("${abspath(path.root)}/data/user-data.pkrtpl.hcl", {
    build_username           = var.build_username
    build_password           = var.build_password
    build_password_encrypted = var.build_password_encrypted
    vm_guest_os_language     = var.vm_guest_os_language
    vm_guest_os_keyboard     = var.vm_guest_os_keyboard
    vm_guest_os_timezone     = var.vm_guest_os_timezone
    network                  = templatefile("${abspath(path.root)}/data/network.pkrtpl.hcl", {
      device  = var.vm_network_device
      ip      = var.vm_ip_address
      netmask = var.vm_ip_netmask
      gateway = var.vm_ip_gateway
      dns     = var.vm_dns_list
    })
    storage                  = templatefile("${abspath(path.root)}/data/storage.pkrtpl.hcl", {
      device     = var.vm_disk_device
      swap       = var.vm_disk_use_swap
      partitions = var.vm_disk_partitions
      lvm        = var.vm_disk_lvm
    })
  })
```

The user-data template will contain something like this:
```
...
  keyboard:
    layout: ${vm_guest_os_keyboard}
${storage}
${network}
  identity:
    hostname: ubuntu-server
    username: ${build_username}
    password: ${build_password_encrypted}
...
```

What do you think?

## Type of Pull Request

- [ ] This is a bugfix. `type/bug`
- [X] This is an enhancement or feature. `type/feature` or `type/enhancement`
- [ ] This is a documentation update. `type/docs`
- [ ] This is a refactoring update. `type/refactor`
- [ ] This is a chore. `type/chore`
- [ ] This is something else.
      Please describe:

## Related to Existing Issues

Closes #262 

## Test and Documentation Coverage

- [ ] Tests have been completed.
- [ ] Documentation has been added or updated.

## Breaking Changes?

- [ ] Yes, there are breaking changes.
- [X] No, there are no breaking changes.

### CODE REVIEW

**Comment by tenthirtyam** [comment]
Sounds great to me @sestegra!

**Comment by sestegra** [comment]
@tenthirtyam Beta version pushed

I have to test every configurations.

**Review by tenthirtyam — COMMENTED** [review]
Thanks, @sestegra - I'll get this reviewed very soon.

File: README.md [hunk 1]
```diff
@@ -799,6 +801,89 @@ Edit the `*.auto.pkrvars.hcl` file in each `builds/<type>/<build>` folder to con
...
   > All `variables.auto.pkrvars.hcl` default to using the [VMware Paravirtual SCSI controller][vmware-pvscsi] and the [VMXNET 3][vmware-vmxnet3] network card device types.
 
+##### Network Variables (Optional)
+
+Edit the `config/network.pkrvars.hcl` file to configure a static IP address:
+
+- IPv4 address, netmask, and gateway.
+- DNS list.
+
+Without any configuration, the network is set using DHCP protocol.
```
Comment (tenthirtyam) [inline README.md:811]:
```suggestion
By default, the network is set to use DHCP for its configuration.
```

File: README.md [hunk 2]
```diff
@@ -799,6 +801,89 @@ Edit the `*.auto.pkrvars.hcl` file in each `builds/<type>/<build>` folder to con
...
+```hcl
+vm_ip_address = "172.16.100.192"
+vm_ip_netmask = 24
+vm_ip_gateway = "172.16.100.1"
+vm_dns_list   = [ "172.16.11.4", "172.16.11.5" ]
+```
+
+> **Note**
+>
+> Static IP assignement is only supported by Linux (except SUSE) for now.
```
Comment (tenthirtyam) [inline README.md:824]:
```suggestion
> Static IP assignment is only supported for the Linux machine images (except SUSE).
```

File: README.md [hunk 3]
```diff
@@ -799,6 +801,89 @@ Edit the `*.auto.pkrvars.hcl` file in each `builds/<type>/<build>` folder to con
...
+    },
+    volume_group = "",
+  },
+]
+vm_disk_lvm = []
+```
+
+> **Note**
+>
+> Storage settings is only supported by Linux (except Photon and SUSE for now).
```
Comment (tenthirtyam) [inline README.md:885]:
```suggestion
> Storage settings are only supported for Linux machine images (except Photon OS and SUSE).
```

Comment (tenthirtyam) [reply README.md:885]: Note: This should be possible in Photon OS 5.0 (Beta.) The installer was updated last year.

File: builds/network.pkrvars.hcl.example [hunk 4]
```diff
@@ -0,0 +1,11 @@
+/*
+    DESCRIPTION:
+    Network variables used for all builds.
+    - Variables are passed to and used by guest operating system configuration files (e.g., ks.cfg, autounattend.xml).
```
Comment (tenthirtyam) [inline builds/network.pkrvars.hcl.example:4]:
```suggestion
    - Variables are passed to and used by guest operating system configuration files (e.g., ks.cfg).
```

File: builds/storage.pkrvars.hcl.example [hunk 5]
```diff
@@ -0,0 +1,153 @@
+/*
+    DESCRIPTION:
+    Storage variables used for all builds.
+    - Variables are passed to and used by guest operating system configuration files (e.g., ks.cfg, autounattend.xml).
```
Comment (tenthirtyam) [inline builds/storage.pkrvars.hcl.example:4]:
```suggestion
    - Variables are passed to and used by guest operating system configuration files (e.g., ks.cfg).
```

File: README.md (hunk 1 above)
Comment (sestegra) [reply README.md:811]: Done

File: README.md (hunk 3 above)
Comment (sestegra) [reply README.md:885]: Done

File: README.md (hunk 2 above)
Comment (sestegra) [reply README.md:824]: Done

File: builds/network.pkrvars.hcl.example (hunk 4 above)
Comment (sestegra) [reply builds/network.pkrvars.hcl.example:4]: Done

File: builds/storage.pkrvars.hcl.example (hunk 5 above)
Comment (sestegra) [reply builds/storage.pkrvars.hcl.example:4]: Done

**Review by tenthirtyam — COMMENTED** [review]
Hi @sestegra 👋 -

I was looking this over and doing some testing this afternoon and wanted to pose a question or two based on the current structure - but first, I should note that from the initial tests, this woeks very nicely.

The concerns or questions that I have are:

- What if you want to have a different storage structure/sizing for each Linux distribution? The current implementation expects the same for all - _should it be unique per build?_
- What if you want to run >1 build at the same time? There would be an issue with the use of a single IP for all builds - _should it be unique per build_?
- In the future, if this is supported for Windows Server images the examples build up to four images and there would not be enough IPs - _should it be unique per build_?
- The storage settings would be different for Linux and Windows (_obviously, Ryan1_) so when adding support for Windows this would need to be refactored since many (most?) of the variables would not be applicable for both - should this be seperated?

Let me know what you think.

**Ryan Johnson**
Senior Staff Solutions Architect | Product Engineering @ VMware, Inc.

**Comment by sestegra** [comment]
Hi @tenthirtyam 

First, I suggest to move all vars files into `config` folder to completely separated code and configuration.
This will ease the `git pull` to update the code only.

See this commit https://github.com/sestegra/packer-examples-for-vsphere/commit/f34601bce7cf07c65df129ffab3d7ad87f1f1e54
- Rename `builds/*/*/*.auto.pkrvars.hcl` to `builds/*/*/*.pkrvars.hcl.example`
- Update `config.sh` to copy all vars files into `config` folder (without subfolders to ease the multiple `config` folders management)
- Update `build.tmpl` to use the copy of the legacy `auto.pkvars.hcl` file as the last variables file. By doing so, you can override global variables from previous files

My suggestion can answer the 2 first questions:
- for common configuration (or sequential build), use `storage` and `network` vars file
- for specific configuration, define variables in the legacy `auto.pkvars.hcl` file that is now located in `config` folder.

For the third question, I have to think about it.

For the last question, I suggest to rename `storage` file into `linux-storage`. Then, as `build.yaml` support specific variables file per build, we can manage future storage management on Windows (`windows-storage`) or BSD.

**Comment by tenthirtyam** [comment]
Thanks for the update, @sestegra - this sounds like a plan to me. Let me know what you'd like to do for the "next steps" and we'll get t pulled into the project. It will certainly be a great improvement.

**Comment by sestegra** [comment]
Next steps;
* [ ] cherry-pick the commit https://github.com/sestegra/packer-examples-for-vsphere/commit/f34601bce7cf07c65df129ffab3d7ad87f1f1e54 to merge into this PR.
* [ ] Rename `storage` file into `linux-storage`
* [ ] Update README.md

**Comment by sestegra** [comment]
Updates done

**Comment by tenthirtyam** [comment]
Hi @sestegra - after (_too_) much time away from the project (_busy in other areas_) I've popped back in to the project to perform some updates. I've rebased this pull requests based on those changes and will look into getting this tested in my environments and ultimately merged very soon. 🚀

**Comment by tenthirtyam** [comment]
Hey @sestegra - you're up next. Do you have time to help resolve the conflicts?

**Comment by sestegra** [comment]
I will have a look

**Comment by sestegra** [comment]
@tenthirtyam Conflicts fixed

**Comment by tenthirtyam** [comment]
Thanks, @sestegra. I'll look for any gaps after the US Thanksgiving week and test. The ones that stood out were Photon 5, Debian 12, and Oracle Linux 8 and 9 which were added for the last release.

**Comment by sestegra** [comment]
@tenthirtyam Happy Thanksgiving week!
I've aligned the new distributions and I've updated the documentation.

**Comment by tenthirtyam** [comment]
Wow! Thank you!!!

**Comment by sestegra** [comment]
@tenthirtyam I fixed conflicts and updated copyright.
The PR is ready to be merged, then I can start working on #694.

**Comment by tenthirtyam** [comment]
I've rebased this to the latest in `develop` and can now start testing again for merge.

**Review by tenthirtyam — APPROVED** [review]
🚀 At long last - tested and approved for merge!

**Pull request merged**